pkg/securitycontextconstraints/seccomp/strategy.go

```diff
@@ -0,0 +1,176 @@
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+const (
+	allowAnyProfile = "*"
+)
+
+// Strategy defines the interface for all seccomp constraint strategies.
+type SeccompStrategy interface {
+	// Generate returns a profile based on constraint rules.
+	Generate(annotations map[string]string, pod *api.Pod) (string, error)
+	// Validate ensures that the specified values fall within the range of the strategy.
+	ValidatePod(pod *api.Pod) field.ErrorList
+	// Validate ensures that the specified values fall within the range of the strategy.
+	ValidateContainer(pod *api.Pod, container *api.Container) field.ErrorList
+}
+
+type strategy struct {
+	allowedProfiles []string
+	// does the strategy allow any profile (wildcard)
+	allowAnyProfile       bool
+	runtimeDefaultAllowed bool
+}
+
+var _ SeccompStrategy = &strategy{}
+
+// NewStrategy creates a new strategy that enforces seccomp profile constraints.
+func NewSeccompStrategy(allowedProfiles []string) SeccompStrategy {
+	allowAny := false
+	allowed := make([]string, 0, len(allowedProfiles))
+	runtimeDefaultAllowed := false
+
+	for _, p := range allowedProfiles {
+		if p == allowAnyProfile {
+			allowAny = true
+			continue
+		}
+		// With the graduation of seccomp to GA we automatically convert
+		// the deprecated seccomp profile `docker/default` to `runtime/default`.
+		// This means that we now have to automatically allow `runtime/default`
+		// if a user specifies `docker/default` and vice versa in an SCC.
+		if p == v1.DeprecatedSeccompProfileDockerDefault || p == v1.SeccompProfileRuntimeDefault {
+			runtimeDefaultAllowed = true
+		}
+		allowed = append(allowed, p)
+	}
+
+	return &strategy{
+		allowedProfiles:       allowed,
+		allowAnyProfile:       allowAny,
+		runtimeDefaultAllowed: runtimeDefaultAllowed,
+	}
+}
+
+// Generate returns a profile based on constraint rules.
+func (s *strategy) Generate(podAnnotations map[string]string, pod *api.Pod) (string, error) {
+	if podAnnotations[api.SeccompPodAnnotationKey] != "" {
+		// Profile already set, nothing to do.
+		return podAnnotations[api.SeccompPodAnnotationKey], nil
+	}
+	if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.SeccompProfile != nil {
+		// Profile field already set, translate to annotation
+		return podutil.SeccompAnnotationForField(pod.Spec.SecurityContext.SeccompProfile), nil
+	}
+
+	// return the first non-wildcard profile
+	if len(s.allowedProfiles) > 0 {
+		return s.allowedProfiles[0], nil
+	}
+
+	return "", nil
+}
+
+// ValidatePod ensures that the specified values on the pod fall within the range
+// of the strategy.
+func (s *strategy) ValidatePod(pod *api.Pod) field.ErrorList {
+	allErrs := field.ErrorList{}
+	podSpecFieldPath := field.NewPath("pod", "metadata", "annotations").Key(api.SeccompPodAnnotationKey)
+	podProfile := pod.Annotations[api.SeccompPodAnnotationKey]
+	// if the annotation is not set, see if the field is set and derive the corresponding annotation value
```

Comment:
why would we honor the annotation instead of using what's in the pod itself?

Reply:
It's a good question, this is copied from upstream. https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/135-seccomp#podsecuritypolicy-enforcement mentions that the actual behavior of the pod might differ based on the version of the node that's running it.